lib/events.js

```diff
@@ -120,47 +126,53 @@ EventEmitter.prototype.emit = function emit(type) {
       listeners[i].apply(this, args);
   }
 
-  if (this.domain && this !== process)
-    this.domain.exit();
+  if (needDomainExit)
+    domain.exit();
 
   return true;
 };
 
 EventEmitter.prototype.addListener = function addListener(type, listener) {
   var m;
+  var events;
+  var existing;
 
-  if (!util.isFunction(listener))
+  if (typeof listener !== 'function')
     throw TypeError('listener must be a function');
 
-  if (!this._events)
-    this._events = {};
-
-  // To avoid recursion in the case that type === "newListener"! Before
-  // adding it to the listeners, first emit "newListener".
-  if (this._events.newListener)
-    this.emit('newListener', type,
-              util.isFunction(listener.listener) ?
-              listener.listener : listener);
+  events = this._events;
+  if (!events)
+    events = this._events = {};
+  else {
+    existing = events[type];
+    // To avoid recursion in the case that type === "newListener"! Before
+    // adding it to the listeners, first emit "newListener".
+    if (events.newListener) {
+      this.emit('newListener', type,
+                typeof listener.listener === 'function' ?
+                listener.listener : listener);
+    }
+  }
 
-  if (!this._events[type])
+  if (!existing)
     // Optimize the case of one listener. Don't need the extra array object.
-    this._events[type] = listener;
-  else if (util.isObject(this._events[type]))
+    existing = events[type] = listener;
+  else if (typeof existing !== 'function')
     // If we've already got an array, just append.
-    this._events[type].push(listener);
+    existing.push(listener);
   else
     // Adding the second element, need to change to array.
-    this._events[type] = [this._events[type], listener];
+    existing = events[type] = [existing, listener];
 
   // Check for listener leak
-  if (util.isObject(this._events[type]) && !this._events[type].warned) {
+  if (typeof existing !== 'function' && !existing.warned) {
```

Comment:
let's switch the order of this conditional logic, no need to do the typeof check which is 2 ops if warned is falsy